astropy/units/quantity.py

```diff
@@ -1452,35 +1467,64 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
-            if precision is None:
+        format_spec = formatter if isinstance(formatter, str) else None
+
+        if format is None:
+            # format_spec overwrites precision
+            if formatter is None and precision is None:
                 # Use default formatting settings
                 return f"{self.value}{self._unitstr:s}"
-            else:
-                # np.array2string properly formats arrays as well as scalars
-                return (
-                    np.array2string(self.value, precision=precision, floatmode="fixed")
-                    + self._unitstr
+
+            if format_spec is not None:
+
+                def formatter(value):
+                    return builtins.format(value, format_spec)
+
+            if callable(formatter):
+                formatter = {"all": formatter}
+
+            return (
+                np.array2string(
+                    self.value,
+                    precision=precision,
+                    floatmode="fixed",
+                    formatter=formatter,
                 )
+                + self._unitstr
+            )
+
+        if format not in formats:
+            raise ValueError(f"Unknown format '{format}'")
 
         # else, for the moment we assume format="latex" or "latex_inline".
 
         # Set the precision if set, otherwise use numpy default
         pops = np.get_printoptions()
-        format_spec = f".{precision if precision is not None else pops['precision']}g"
+        if format_spec is None:
+            format_spec = (
+                f".{precision if precision is not None else pops['precision']}g"
+            )
 
-        def float_formatter(value):
-            return Latex.format_exponential_notation(value, format_spec=format_spec)
+        # Use default formatters
+        if formatter is None or isinstance(formatter, str):
+            if self.dtype.kind == "c":  # Complex default latex formatter
+
+                def formatter(value):
+                    return "({}{}i)".format(
+                        Latex.format_exponential_notation(
+                            value.real, format_spec=format_spec
+                        ),
+                        Latex.format_exponential_notation(
+                            value.imag, format_spec="+" + format_spec
+                        ),
+                    )
 
-        def complex_formatter(value):
-            return "({}{}i)".format(
-                Latex.format_exponential_notation(value.real, format_spec=format_spec),
-                Latex.format_exponential_notation(
-                    value.imag, format_spec="+" + format_spec
-                ),
-            )
+            else:  # Float default latex formatter
+
+                def formatter(value):
+                    return Latex.format_exponential_notation(
+                        value, format_spec=format_spec
+                    )
 
         # The view is needed for the scalar case - self.value might be float.
```

Comment:
Here, you need to do also for the latex branch,
```
        if callable(formatter):
            formatter = {"all": formatter}
```
and then use `formatter=formatter` in the `array2string` call.